### PR TITLE
A possibly better fix for #313

### PR DESCRIPTION
**Pull request by Kethku**
This stores keycode and textinput information in a vector of pairs so that subsequent keycode and textinput events are kept together irrespective of the order in which they are pressed.

This seems to fix the issue of missing key presses but does not always get the ordering correct. This may be a limitation of the sdl keyboard input features.

@j4qfrost any thoughts? Sorry again for prematurely merging the other branch.

### CODE REVIEW

**Comment by j4qfrost** [comment]
So the issue that was blocking my changes is that text would double up. I'm having trouble tracing how the extra text gets deduped.

**Comment by Kethku** [comment]
My branch fixes that problem. I think the doubling would occur when events for both the keycode and the keytext got send separately from eachother. This gets around that by pairing them up. The assumption being that keycode and keytext will always come in pairs.

File: src/window.rs [hunk 1]
```diff
@@ -438,9 +440,31 @@ pub fn ui_loop() {
...
                     keycode = received_keycode;
+
+                    if keytext.is_some() {
+                        keyboard_inputs.push((keycode, keytext));
+                        keycode = None;
+                        keytext = None;
+                    }
+                }
+                Event::TextInput { text, .. } => {
+                    if keytext.is_some() {
```
Comment (j4qfrost) [inline src/window.rs:456]: This is where I'm afraid the double happens.

File: src/window.rs [hunk 2]
```diff
@@ -438,9 +440,31 @@ pub fn ui_loop() {
                     keycode: received_keycode,
                     ..
                 } => {
+                    if received_keycode.is_some() {
```
Comment (j4qfrost) [inline src/window.rs:443]: Also doubling up here?

File: src/window.rs (hunk 1 above)
Comment (Kethku) [reply src/window.rs:456]: id be interested in seeing how it works on non windows platforms. It doesn't double up on windows

File: src/window.rs (hunk 2 above)
Comment (Kethku) [reply src/window.rs:443]: this logic says if there is already a keycode then we can assume that the new keycode is a different key. So we send that keycode to the vector to get processed and then store the new keycode in the temp variable to be paired with a textinput if it becomes available.

File: src/window.rs (hunk 1 above)
Comment (j4qfrost) [reply src/window.rs:456]: Grabbing now.

File: src/window.rs (hunk 2 above)
Comment (Kethku) [reply src/window.rs:443]: So in practice no doubling up happens

Comment (j4qfrost) [reply src/window.rs:443]: Wait I think there's a typo then. You want to check keycode and not received_keycode right?

Comment (Kethku) [reply src/window.rs:443]: I'll add some comments and double check

File: src/window.rs (hunk 1 above)
Comment (j4qfrost) [reply src/window.rs:456]: There isn't any double up.

File: src/window.rs (hunk 2 above)
Comment (Kethku) [reply src/window.rs:443]: yup you're totally right. Doh

Comment (Kethku) [reply src/window.rs]: Fixed and pushed. I also reformatted it slightly for dryness and readability

**Comment by j4qfrost** [comment]
Looks good.

**Comment by Kethku** [comment]
I'm going to test it just a bit more, (the ordering thing is still bugging me) but then I'll probably merge

**Pull request merged**